### PR TITLE
201 aci sftp files

### PR DESCRIPTION
**Pull request by bhummerstone**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

* Initial PR
*
*

### CODE REVIEW

File: 201-aci-sftp-files/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+        "sftpContainerName": "sftp",
+        "sftpContainerGroupName": "sftp-group",
+        "sftpContainerImage": "atmoz/sftp:latest",
+        "sftpEnvVariable": "[concat(parameters('sftpUser'), ':', parameters('sftpPassword'), ':1001')]",
+        "storageAccountName": "[concat(parameters('storageAccountPrefix'), uniqueString(resourceGroup().id))]",
+        "createFileShareTemplateUri": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/201-aci-sftp-files/nestedtemplates/createFileShare.json"
+    },
+    "resources": [
+        {
+            "type": "Microsoft.Resources/deployments",
```
Comment (bmoore-msft) [inline 201-aci-sftp-files/azuredeploy.json]: what's the purpose of this resource?

File: 201-aci-sftp-files/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+            }
+        }
+    },
+    "variables": {
+        "cliContainerName": "create-share",
+        "cliContainerGroupName": "create-share-group",
+        "cliContainerImage": "microsoft/azure-cli:latest",
+        "sftpContainerName": "sftp",
+        "sftpContainerGroupName": "sftp-group",
+        "sftpContainerImage": "atmoz/sftp:latest",
```
Comment (bmoore-msft) [inline 201-aci-sftp-files/azuredeploy.json]: who owns this image?

File: 201-aci-sftp-files/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,163 @@
+{
+    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+    "contentVersion": "1.0.0.0",
+    "parameters": {
+        "storageAccountPrefix": {
```
Comment (bmoore-msft) [inline 201-aci-sftp-files/azuredeploy.json]: move this to a variable

File: 201-aci-sftp-files/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        "fileShareName": {
+            "type": "string",
+            "defaultValue": "sftpfileshare",
+            "metadata": {
+                "description": "Name of file share to be created"
+            }
+        },
+        "sftpUser": {
+            "type": "string",
+            "defaultValue": "sftp",
```
Comment (bmoore-msft) [inline 201-aci-sftp-files/azuredeploy.json]: remove the defaultValue - increases surface area

File: 201-aci-sftp-files/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+    "variables": {
+        "cliContainerName": "create-share",
+        "cliContainerGroupName": "create-share-group",
+        "cliContainerImage": "microsoft/azure-cli:latest",
+        "sftpContainerName": "sftp",
+        "sftpContainerGroupName": "sftp-group",
+        "sftpContainerImage": "atmoz/sftp:latest",
+        "sftpEnvVariable": "[concat(parameters('sftpUser'), ':', parameters('sftpPassword'), ':1001')]",
+        "storageAccountName": "[concat(parameters('storageAccountPrefix'), uniqueString(resourceGroup().id))]",
+        "createFileShareTemplateUri": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/201-aci-sftp-files/nestedtemplates/createFileShare.json"
```
Comment (bmoore-msft) [inline 201-aci-sftp-files/azuredeploy.json]: See [BP guide](https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#deployment-artifacts-nested-templates-scripts) for how to handle artifacts

File: 201-aci-sftp-files/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                "mode": "Incremental",
+                "template": {
+                    "$schema": "https://schema.management.azure.com/schemas/2015-01-01/deploymentTemplate.json#",
+                    "contentVersion": "1.0.0.0",
+                    "resources": []
+                }
+            }
+        },
+        {
+            "type": "Microsoft.Resources/deployments",
```
Comment (bmoore-msft) [inline 201-aci-sftp-files/azuredeploy.json]: remove the nesting for this and just put the storageAccount in this template

File: 201-aci-sftp-files/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+                        "value": "[parameters('fileShareName')]"
+                    }
+                }
+            }
+        },
+        {
+            "type": "Microsoft.ContainerInstance/containerGroups",
+            "name": "[variables('sftpContainerGroupName')]",
+            "apiVersion": "2018-04-01",
+            "location": "[resourceGroup().location]",
```
Comment (bmoore-msft) [inline 201-aci-sftp-files/azuredeploy.json]: Add location parameter - see: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#parameters

File: 201-aci-sftp-files/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+                        }
+                    }
+                ]
+            }
+        }
+    ],
+    "outputs": {
+        "containerIPv4Address": {
+            "type": "string",
+            "value": "[reference(concat('Microsoft.ContainerInstance/containerGroups/', variables('sftpContainerGroupName'))).ipAddress.ip]"
```
Comment (bmoore-msft) [inline 201-aci-sftp-files/azuredeploy.json]: remove concat and use resourceId()

File: 201-aci-sftp-files/nestedtemplates/createFileShare.json [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        "cliContainerName": "create-share",
+        "cliContainerGroupName": "create-share-group",
+        "cliContainerImage": "microsoft/azure-cli:latest"
+    },
+    "resources": [
+        {
+            "type": "Microsoft.Storage/storageAccounts",
+            "name": "[parameters('storageAccountName')]",
+            "apiVersion": "2018-02-01",
+            "location": "[resourceGroup().location]",
```
Comment (bmoore-msft) [inline 201-aci-sftp-files/nestedtemplates/createFileShare.json]: Add location parameter - see: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#parameters

File: 201-aci-sftp-files/azuredeploy.json (hunk 2 above)
Comment (bhummerstone) [reply 201-aci-sftp-files/azuredeploy.json]: Added attribution in the readme

File: 201-aci-sftp-files/azuredeploy.json (hunk 3 above)
Comment (bhummerstone) [reply 201-aci-sftp-files/azuredeploy.json]: Done

File: 201-aci-sftp-files/azuredeploy.json (hunk 4 above)
Comment (bhummerstone) [reply 201-aci-sftp-files/azuredeploy.json]: Done

File: 201-aci-sftp-files/azuredeploy.json (hunk 5 above)
Comment (bhummerstone) [reply 201-aci-sftp-files/azuredeploy.json]: Removed nesting, so no longer required

File: 201-aci-sftp-files/azuredeploy.json (hunk 6 above)
Comment (bhummerstone) [reply 201-aci-sftp-files/azuredeploy.json]: Done

File: 201-aci-sftp-files/azuredeploy.json (hunk 7 above)
Comment (bhummerstone) [reply 201-aci-sftp-files/azuredeploy.json]: Done

File: 201-aci-sftp-files/azuredeploy.json (hunk 8 above)
Comment (bhummerstone) [reply 201-aci-sftp-files/azuredeploy.json]: Done

File: 201-aci-sftp-files/nestedtemplates/createFileShare.json (hunk 9 above)
Comment (bhummerstone) [reply 201-aci-sftp-files/nestedtemplates/createFileShare.json]: Done

File: 201-aci-sftp-files/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,195 @@
...
+                "ipAddress": {
+                    "type": "Public",
+                    "ports": [
+                        {
+                            "protocol": "TCP",
+                            "port": 22
+                        }
+                    ]
+                },
+                "restartPolicy": "OnFailure",
```
Comment (wenwu449) [inline 201-aci-sftp-files/azuredeploy.json]: If this is going to host the sftp service, ["Always"](https://docs.microsoft.com/en-us/azure/container-instances/container-instances-restart-policy#container-restart-policy)  may be helpful.

Comment (bhummerstone) [reply 201-aci-sftp-files/azuredeploy.json]: The idea is that the container only needs to be available for the life of the transfer, so the restartPolicy isn't as important imo :)

**Pull request merged**